### ISSUE TITLE
Unsafe precondition violations in Rust 1.78.0

### ISSUE THREAD

**Issue by LPGhatguy** [lebedec/libfmod#13]
Rust 1.78.0 started correctly surfacing some unsafe assertions within the standard library as panics. Since upgrading to Rust 1.78.0, I've started hitting panics when calling some libfmod functions due to unaligned pointers.

I hit this issue calling `System::get_advanced_settings`, which calls `AdvancedSettings::try_from`, which uses the [`to_vec` macro](https://github.com/lebedec/libfmod/blob/4339ed1331177d5763ba25c4ea0f9a78a33aab70/src/lib.rs#L115-L126), that invokes `slice::from_raw_parts` on a pointer that Rust is asserting is invalid.

The error that Rust panics with in this case is:

> unsafe precondition(s) violated: slice::from_raw_parts requires the pointer to be aligned and non-null, and the total size of the slice not to exceed `isize::MAX`

**Comment by lebedec** [comment]
Hi! If you have any ideas on how to change the macro, it will help a lot.

**Comment by lebedec** [comment]
docs https://blog.rust-lang.org/2024/05/02/Rust-1.78.0.html#asserting-unsafe-preconditions

**Comment by LPGhatguy** [comment]
To unblock my team, I [applied a patch to my fork of libfmod](https://github.com/SecondHalfGames/libfmod/commit/3003a3e4c83c099cd92f0dea27ded08e8424519e) which stops the panic for now and doesn't _seem_ to break anything in our game. It is not a good patch.

> Hi! If you have any ideas on how to change the macro, it will help a lot.

Originally I thought this panic would only happen on unaligned pointers, but looking at [the implementation of the checks](https://doc.rust-lang.org/1.78.0/src/core/slice/raw.rs.html#95) and reading the docs more closely, there are some other possibilities!

In decreasing likelihood, here are my ideas and suggestions:

## 1: The pointer could be null
For `ASIOChannelList`, one of the fields in the `FMOD_ADVANCEDSETTINGS` struct, the FMOD docs say:

> Read only list of strings representing ASIO channel names ([UTF-8 string](https://fmod.com/docs/2.02/api/glossary.html#string-format)), count is defined by ASIONumChannels. Only valid after [System::init](https://fmod.com/docs/2.02/api/core-api-system.html#system_init).

I am calling `get_advanced_settings` before `System::init` because of the specific setting I'm changing, so it makes sense that this pointer might be null. Handling null in the macro would be an easy fix.

## 2: The pointer could be unaligned
It's possible that the pointer is not null, but doesn't meet the minimum alignment of the FFI type. If this is the case, this is an upstream bug that we should report, because that's UB in C, C++, and Rust.

This could be worked around by pre-allocating a `Vec`, using [`ptr::read_unaligned`](https://doc.rust-lang.org/std/ptr/fn.read_unaligned.html) to copy each element into it, then continuing with the rest of the `into_iter().collect()` process.

## 3: The pointer could be invalid
It's very unlikely, but possible that FMOD is giving back and uninitialized pointer and it's pointing to some random address. If this is the case, I don't think there's much we can do from the Rust side. This would be an upstream bug to report.

**Comment by lebedec** [comment]
Thanks for the ideas. It will take me some time to make fix

**Comment by lebedec** [comment]
I looked at usages of `to_vec` macro. There is always a "length" argument that is responsible for correct usage of pointer. So I decided that proper solution would be just check this argument for zero and create empty `Vec` if needed.

Cases with incorrect pointer I ignored (including case with null pointer but not zero "length"). As you said, in these cases it’s most likely a FMOD bug that makes no sense to hide (handle quietly).

Fixed in libfmod `2.206.6`

```rust
#[test]
fn test_system_advanced_settings_before_init() -> Result<(), Error> {
    let system = System::create()?;
    let settings = system.get_advanced_settings()?;
    println!("channels: {:?}", settings.asio_channel_list.is_empty());
}
```

**Comment by lebedec** [comment]
@LPGhatguy thanks for help!
